### PR TITLE
add apache solr sample app

### PR DESCRIPTION
**Pull request by JonathanWamsley**
This Sample app uses multipass to create an Apache Solr cluster. A Makefile is used which has commands to create the apache solr, add load gen, view prometheus metrics and to set up a grafana agent using a static grafana-agent.yaml file to send metrics/logs to grafana cloud.

### CODE REVIEW

File: apache-solr/README.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+- Basic understanding of Solr, ZooKeeper, and Grafana.
+
+## Versions
+Apache Solr: 8.11.2
+ZooKeeper: 3.4.13
+Grafana Agent: 0.39.0-1 for arm64 architecture (adjust as per your architecture)
+
+## Configuration
+
+Before running the setup, ensure to fill out the `grafana-agent-template.yaml` with your Grafana Cloud credentials and endpoints. Replace placeholders like `<your grafana.com password>`, `<your grafana.com username>`, `<your prometheus push endpoint>`, etc., with actual values. Rename the file to `grafana-agent.yaml` after configuration.
```
Comment (rgeyer) [inline apache-solr/README.md:18]: This works fine for an interactive user doing something like acceptance testing, but makes it difficult to automate, which is an ultimate goal.

We haven't defined a repeatable/shared way to do this yet (some shared libs/scripts/patterns are forthcoming), but in all of the sample apps I've written, I'm using jinja templates and a Docker container execution to render those templates with appropriate variables. I.E. You can ask the user on the shell for the variables, then feed them into jinja templates.

I'm open to other/better solutions, and I would like to make those steps simpler/reusable in this repo.

Example from the MySQL Sample App: https://github.com/grafana/integration-sample-apps/blob/c0916d21dae6b9d130af99e00226e0954609b450/mysql/Makefile#L13-L41

File: apache-solr/scripts/load_generation.sh [hunk 2]
```diff
@@ -0,0 +1,50 @@
+#!/bin/bash
```
Comment (rgeyer) [inline apache-solr/scripts/load_generation.sh:1]: A bit of a thought exercise, with no real call to action.

Would this be better/worse/different if we used k6 to run these tests?

**Review by rgeyer — COMMENTED** [review]
Looks pretty good. I've asked a few questions, and made some small suggestions. I'd like to have some discourse on those before approving and merging.

Great work tho, particularly given how little direction and prior art there is for these sample apps.

File: apache-solr/Makefile [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    done
+    # Clean up
+	rm vms_ips.txt
+
+load-test:
+	multipass transfer scripts/load_generation.sh $(VM_PREFIX)-1:/home/ubuntu/load_generation.sh; \
+    multipass exec $(VM_PREFIX)-1 -- bash -c "chmod +x /home/ubuntu/load_generation.sh && nohup /home/ubuntu/load_generation.sh > load_gen.log 2>&1 &"
+
+fetch-prometheus-metrics:
+	multipass exec $(VM_PREFIX)-1 -- bash -c "curl http://localhost:9854/metrics" > prometheus_metrics
```
[the remark this answers is not in the document]
Comment (JonathanWamsley) [reply apache-solr/Makefile:39]: noted and shared for future sample apps 👍

File: apache-solr/Makefile [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        multipass transfer grafana-agent.yaml $(VM_PREFIX)-$$i:/home/ubuntu/grafana-agent.yaml; \
+        multipass transfer scripts/install_grafana_agent.sh $(VM_PREFIX)-$$i:/home/ubuntu/install_grafana_agent.sh; \
+        multipass exec $(VM_PREFIX)-$$i -- bash -c "chmod +x /home/ubuntu/install_grafana_agent.sh && /home/ubuntu/install_grafana_agent.sh"; \
+    done
+
+clean:
+    # Delete VMs
+	for i in 1 2 3; do \
+        multipass delete $(VM_PREFIX)-$$i && multipass purge; \
+    done
```
[the remark this answers is not in the document]
Comment (JonathanWamsley) [reply apache-solr/Makefile]: Same, though we noticed `clean` is already used, so updated to `stop` to stop, delete and purge the running cluster.

File: apache-solr/README.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
+# Apache Solr and ZooKeeper Cluster Sample App
+
+This project sets up a cluster of Apache Solr and ZooKeeper instances, integrated with Grafana agent for metric and log collection. It uses Multipass to create virtual machines where Apache Solr and ZooKeeper are deployed and configured. Additionally, a Grafana Agent is installed for metrics and logs collection and forwarding.
```
[the remark this answers is not in the document]
Comment (JonathanWamsley) [reply apache-solr/README.md]: updated docs and sorta standardized parts with with the openLDAP version.  This helps new users get started.

File: apache-solr/README.md (hunk 1 above)
Comment (JonathanWamsley) [reply apache-solr/README.md]: Added jinja templates for the grafana agent section

File: apache-solr/scripts/load_generation.sh (hunk 2 above)
Comment (JonathanWamsley) [reply apache-solr/scripts/load_generation.sh:1]: cool, I'll play with this more in detail. We have a free day coming up, so I can spend some time trying this out for future sample apps.

File: apache-solr/jinja/templates/cloud-init-template.yaml [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+      integrations:
+        node_exporter:
+          enabled: true
+          relabel_configs:
+          - replacement: hostname
+            target_label: instance
+        prometheus_remote_write:
+        - basic_auth:
+            password: "{{ prom_pass }}"
+            username: "{{ prom_user }}"
```
Comment (rgeyer) [inline apache-solr/jinja/templates/cloud-init-template.yaml:17]: This should be conditional. If the prom_user and prom_pass are unset, then it implies that no authentication is required.

An example from the MySQL [jinja template.](https://github.com/grafana/integration-sample-apps/blob/c0916d21dae6b9d130af99e00226e0954609b450/mysql/jinja/templates/cloud-init-template.yaml#L34-L39)

File: apache-solr/Makefile [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+setup-grafana-agent:
+	for i in 1 2 3; do \
+        multipass transfer scripts/install_grafana_agent.sh $(VM_PREFIX)-$$i:/home/ubuntu/install_grafana_agent.sh; \
+        multipass exec $(VM_PREFIX)-$$i -- bash -c "chmod +x /home/ubuntu/install_grafana_agent.sh && /home/ubuntu/install_grafana_agent.sh"; \
+	done
+
+stop:
+	# Delete VMs
+	for i in 1 2 3; do \
+        multipass delete $(VM_PREFIX)-$$i && multipass purge; \
```
[the remark this answers is not in the document]
Comment (JonathanWamsley) [reply apache-solr/Makefile]: updated 👍

File: apache-solr/README.md [hunk 8]
```diff
@@ -0,0 +1,107 @@
+# Apache Solr and ZooKeeper Cluster Sample App
+
+This sample application creates 3 VMs each with an Apache Solr and ZooKeeper instance, integrated with Grafana Agent for metric and log collection. This sample app utilizes cloud-init and Make commands to facilitate the setup, configuration, and monitoring of Apache Solr and ZooKeeper instances.
+
+*Note*: In order to have a working apache-solr-zookeeper-instances, additional memory and cpu resources are allocated to each VM.
```
[the remark this answers is not in the document]
Comment (JonathanWamsley) [reply apache-solr/README.md:5]: Glad you like it 😁

File: apache-solr/grafana-agent-template.yaml [hunk 9]
```diff
@@ -0,0 +1,56 @@
+integrations:
```
[the remark this answers is not in the document]
Comment (JonathanWamsley) [reply apache-solr/grafana-agent-template.yaml]: removed

File: apache-solr/jinja/templates/cloud-init-template.yaml (hunk 6 above)
Comment (JonathanWamsley) [reply apache-solr/jinja/templates/cloud-init-template.yaml]: updated

File: apache-solr/jinja/templates/cloud-init-template.yaml [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+          url: "{{ prom_url }}"
+      logs:
+        configs:
+        - name: integrations/apache-solr
+          clients:
+            - url: "{{ loki_url }}"
```
[the remark this answers is not in the document]
Comment (JonathanWamsley) [reply apache-solr/jinja/templates/cloud-init-template.yaml:25]: added conditional auth

**Review by rgeyer — APPROVED** [review]
🚀

File: apache-solr/Makefile [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+stop:
+	# Delete VMs
+	for i in 1 2 3; do \
+        multipass delete $(VM_PREFIX)-$$i; \
+    done
+	multipass purge
+
+render-grafana-agent-config:
+	@docker run --rm -it -v $(shell pwd)/jinja/templates:/templates -v $(shell pwd)/jinja/variables:/variables dinutac/jinja2docker:latest /templates/cloud-init-template.yaml /variables/cloud-init.yaml --format=yaml > cloud-init.yaml
```
Comment (Dasomeone) [inline apache-solr/Makefile:55]: This command defaults to amd64, can you update this to take the host platform into account? (or add a note on the readme on supported platforms/known issues)

```
Status: Downloaded newer image for dinutac/jinja2docker:latest
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
```

Comment (Dasomeone) [reply apache-solr/Makefile:55]: I had a look and it seems `dinutac/jinja2docker:latest` only supports amd64 so we'll need a platform support section in the readme

**Review by Dasomeone — COMMENTED** [review]
A few comments from my side after trying to set it up, but overall my thoughts alight completely with Ryan's here. Looks great, neat config, etc etc and I really like how clean it is.

Unfortunately couldn't test it on arm64 as mentioned so we do need a platform support / known issues section high up on the readme, but I'll go and double check it on my amd64 machine in a sec so more comments soon:tm:

File: apache-solr/Makefile (hunk 11 above)
Comment (rgeyer) [reply apache-solr/Makefile:55]: This was a rather random choice on my part. We can find a different docker image, or create our own for jinja templating. 🤔

File: apache-solr/README.md [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+To get started with the Apache Solr and ZooKeeper cluster along with monitoring tools, follow these steps:
+
+1. **Clone the Repository**: 
+   ```sh
+   git clone https://github.com/grafana/integration-sample-apps.git
+   cd integration-sample-apps/apache-solr
+   ```
+
+2. **Set Up Default Config**: 
+   Execute `make defaultconfig` to create a template file with default configuration variables. Modify `jinja/variables/cloud-init.yaml` to connect the Grafana agent to an external Prometheus compatible TSDB and/or Loki server.
```
[the remark this answers is not in the document]
Comment (JonathanWamsley) [reply apache-solr/README.md:29]: added in the default cloud-init.yaml file

File: apache-solr/Makefile [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+	@echo "prom_user: your_prometheus_username" >> jinja/variables/cloud-init.yaml
+	@echo "prom_url: http://your-prometheus-instance:9090/api/v1/push" >> jinja/variables/cloud-init.yaml
+	@echo "solr_cluster_name: solr-cluster" >> jinja/variables/cloud-init.yaml
+	@echo "solr_host: localhost" >> jinja/variables/cloud-init.yaml
+	@echo "solr_port: 9854" >> jinja/variables/cloud-init.yaml
+	@echo "solr_log_path: /var/solr/logs/*.log" >> jinja/variables/cloud-init.yaml
+	@echo "instance_name: solr-instance" >> jinja/variables/cloud-init.yaml
+	@echo "loki_url: http://your-loki-instance:3100/loki/api/v1/push" >> jinja/variables/cloud-init.yaml
+	@echo "loki_user: your_loki_username" >> jinja/variables/cloud-init.yaml
+	@echo "loki_pass: your_loki_password" >> jinja/variables/cloud-init.yaml
```
[the remark this answers is not in the document]
Comment (JonathanWamsley) [reply apache-solr/Makefile:68]: Added in the default cloud-init.yaml

File: apache-solr/README.md [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+
+3. **Render Cloud-init Configuration**: 
+   Run `make render-config` to generate the `cloud-init.yaml` file based on your configuration.
```
[the remark this answers is not in the document]
Comment (JonathanWamsley) [reply apache-solr/README.md:32]: renamed to render-config

File: apache-solr/Makefile (hunk 11 above)
Comment (JonathanWamsley) [reply apache-solr/Makefile]: added docs saying it only supports amd64

File: apache-solr/Makefile [hunk 15]
```diff
@@ -0,0 +1,71 @@
+VM_PREFIX=apache-solr-zookeeper-instance
+
+.PHONY: run create-vms setup-zookeeper setup-solr load-test fetch-prometheus-metrics setup-grafana-agent clean render-grafana-agent-config defaultconfig
+
+run: create-vms setup-zookeeper setup-solr load-test fetch-prometheus-metrics setup-grafana-agent
+	@echo "VMs and services configured."
```
[the remark this answers is not in the document]
Comment (JonathanWamsley) [reply apache-solr/Makefile:6]: hmm, I added in a `DEBIAN_FRONTEND=noninteractive` before and I stopped getting those interruptions. I added another option `-o Dpkg::Options::="--force-confnew"` when setting up zookeeper. Let me know if this works for you.

[the remark this answers is not in the document]
Comment (Dasomeone) [reply apache-solr/Makefile:6]: Hmmm,  not quite unfortunately. @rgeyer did you encounter anything similar to this after Jon's changes (or even before?) 
Curious to see if this is just me missing something in my setup

image

**Review by Dasomeone — COMMENTED** [review]
Thanks for addressing the comments Jon! Just a few more, but if Ryan isn't having issues then I'm happy to merge on his approval

**Review by Dasomeone — APPROVED** [review]
LGTM, and just for records sake, Jon showed me the commands working without halting for log inspection like on my system, so I'm happy with it as is 😄

**Pull request merged**